pkg/planner/core/rule_predicate_simplification.go

```diff
@@ -147,6 +190,116 @@ func applyPredicateSimplification(sctx base.PlanContext, predicates []expression
 	return newValues
 }
 
+// Check for constant false condition.
+func unsatisfiableExpression(ctx base.PlanContext, p expression.Expression) bool {
+	if constExpr, ok := p.(*expression.Constant); ok {
+		if b, err := constExpr.Value.ToBool(ctx.GetSessionVars().StmtCtx.TypeCtx()); err == nil && b == 0 {
+			return true
+		}
+	}
+	return false
+}
+
+func unsatisfiable(ctx base.PlanContext, p1, p2 expression.Expression) bool {
+	var equalPred expression.Expression
+	var otherPred expression.Expression
+	col1, p1Type := findPredicateType(p1)
+	col2, p2Type := findPredicateType(p2)
+	if col1 != col2 || col1 == nil {
+		return false
+	}
+	if p1Type == equalPredicate {
+		equalPred = p1
+		otherPred = p2
+	} else if p2Type == equalPredicate {
+		equalPred = p2
+		otherPred = p1
+	}
+	if equalPred == nil || otherPred == nil {
+		return false
+	}
+	// Copy constant from equal predicate into other predicate.
+	equalValue := equalPred.(*expression.ScalarFunction)
+	otherValue := otherPred.(*expression.ScalarFunction)
+	newPred := expression.NewFunctionInternal(ctx.GetExprCtx(), otherValue.FuncName.L, otherValue.RetType, equalValue.GetArgs()[1], otherValue.GetArgs()[1])
+	newPredList := make([]expression.Expression, 0, 1)
+	newPredList = append(newPredList, newPred)
+	newPredList = expression.PropagateConstant(ctx.GetExprCtx(), newPredList)
+	return unsatisfiableExpression(ctx, newPredList[0])
+}
+func comparisonPred(predType predicateType) predicateType {
+	if predType == equalPredicate || predType == lessThanPredicate ||
+		predType == greaterThanPredicate || predType == lessThanOrEqualPredicate ||
+		predType == greaterThanOrEqualPredicate {
+		return scalarPredicate
+	}
+	return predType
+}
+
+// updateOrPredicate simplifies OR predicates by dropping OR predicates if they are empty.
+// It is applied for this pattern: P AND (P1 OR P2 ... OR Pn)
+// Pi is removed if P & Pi is false/empty.
+func updateOrPredicate(ctx base.PlanContext, orPredicateList expression.Expression, scalarPredicatePtr expression.Expression) expression.Expression {
+	_, orPredicateType := findPredicateType(orPredicateList)
+	_, scalarPredicateType := findPredicateType(scalarPredicatePtr)
+	scalarPredicateType = comparisonPred(scalarPredicateType)
+	if orPredicateType != orPredicate || scalarPredicateType != scalarPredicate {
+		return orPredicateList
+	}
+	v := orPredicateList.(*expression.ScalarFunction)
+	firstCondition := v.GetArgs()[0]
+	secondCondition := v.GetArgs()[1]
+	_, firstConditionType := findPredicateType(firstCondition)
```

Comment:
This section seems to contain identical code segments. Could a single function, or even an anonymous function, be used to simplify the code?

Reply:
252-267L

Reply:
Agree the code is repeated for argument 0 and 1 and I considered factoring those. But, the common code is recursive and I would rather avoid this indirect recursion. Let me see if I can refactor and keep the direct recursion.

Reply:
Actually, it's not a big issue. Sometimes Golang will inline this part of the code by itself.